Workaround for buttonless Controllers

bonkzwonil opened the pull request:
Some controllers as the Thrustmaster T-Rudder have no buttons and thus throw an KeyError.
Test buttons first and insert an empty set for the buttons fixes not only this issue but also makes evdevremapkeys and excellent tool to add those controllers to steam/wine/proton, as thats ignoring controllers without buttons :)

To make those work just create an empty mapping like this:

```
devices:
  - input_name: "Thrustmaster T-Rudder"
    output_name: "T-Rudder"
    remappings: {}
```

philipl commented:
This looks reasonable. Can you please update it to fix the conflict - it's happening due to the new standardised formatting.

philipl commented:
Also, please remove the `print(caps)` that snuck back in, and can you explain the hard-coded button?

bonkzwonil commented:
I did that and also made it configurable instead of hardcoded.

Explaination for the docs:
wine is picky and ignores devices when they have no buttons, I found it needs at least 2 buttons (ecodes.BTN_A and BTN_B)
Also it looks into the product/vendor ids to map them to xinput/dinput

What now works for my setup for DCS:
I have Thrustmaster Warthog setup with some pedals (Thats the one with no buttons)

My config:

```
devices:
  - input_name: "Thustmaster Joystick - HOTAS Warthog"
    output_name: "Joystick - HOTAS Warthog"
    remappings: {}
  - input_name: "Thrustmaster Throttle - HOTAS Warthog"
    output_name: "Throttle - HOTAS Warthog"
    remappings: {}
  - input_name: "Thrustmaster T-Rudder"
    output_name: "T-Rudder"
    product_id: 0x1234
    dummy_buttons: [304, 305]
    remappings: {}

```

This works now (with default product_id it gets ignored). I also added the thrupassing of input ids/product, because some games (like DCS) depend on orginal names and ids (thats why I also added the outputname)

In wine it may be need to get whitelisted as well via `wine control` 

![image](https://github.com/user-attachments/assets/76ccee4d-ac93-4e43-b213-cce009b2f5b7)


But often its not necessary

bonkzwonil commented:
nb: the fixed/passed thru productid was also necessary because otherwise DCS creates a new profile every run (UInput randomizes them otherwise)

philipl commented on evdevremapkeys/evdevremapkeys.py:
```diff
@@ -328,8 +334,15 @@ def flatmap(lst):
-    output = UInput(caps, name=device["output_name"])
-    print("Registered: %s, %s, %s" % (input.name, input.path, input.phys), flush=True)
+
+    output = UInput(caps, 
+    name=device['output_name'],
+    vendor=input.info.vendor if not 'vendor_id' in device else device['vendor_id'],        
+    product=input.info.product if not 'product_id' in device else device['product_id'],      
+    version=input.info.version if not 'version' in device else device['version'],     
+    bustype=0x06 if not 'bustype' in device else device['bustype'],         # USB 0x06, virtual 0x03 etc
+    )
```
Please do this with a dict and **kwargs so that you only set the key/value in the dict if it's set in the device. That way we are not overriding the way UInput calculates default values.

bonkzwonil replied:
alright did that.

However I also did some more changes, to reuse output if the input is reconnecting. (for example due to a shitty usb cable.)
This is now included as well.

If you dont want this I could make a branch to exclude it. But it works quite well.

philipl replied:
Thanks! But please do move the reconnect logic to another PR so I can read it more easily. Much appreciated!

bonkzwonil replied:
alright, Its done.

Will open a new PR for the reuse

this one should be squashed or it will get messy ;)

philipl reviewed (changes requested):
Alright.

Please rebase and make sure the change passes linting and formatting. Also, please add an example to the sample file - that's our primary source of documentation.

Thanks.

philipl commented:
Thanks for updating everything. You still need to rebase onto master before I can merge it as github is reporting conflicts. Also, please update the PR description to reflect the final state, as that will be what is used for the squashed commit.

Thanks again!

philipl commented:
I've pushed a series of changes that implement equivalent functionality.
